### PR TITLE
os/bluestore: fix bluefs migrate command

### PR DESCRIPTION
**Pull request by ifed01**
Fixes: https://tracker.ceph.com/issues/40434
Signed-off-by: Igor Fedotov <ifedotov@suse.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by aclamk** [comment]
@ifed01 assigned or review?

**Comment by ifed01** [comment]
> @ifed01 assigned or review?

@aclamk - sorry, certainly review ;)

**Comment by aclamk** [comment]
@ifed01 
I am unable to replicate problem in the first place.
I tried:
1) 
- Start with SLOW,DB,WAL (regular vstart)
- get rid of DB
-> fsck works, cluster works
2)
- Start with SLOW,DB (hacked vstart to bluestore_block_wal_create=false)
- get rid of DB
-> fsck works, cluster works

**Comment by ifed01** [comment]
> @ifed01
> I am unable to replicate problem in the first place.
> I tried:
> 1)
> 
> * Start with SLOW,DB,WAL (regular vstart)
> * get rid of DB
>   -> fsck works, cluster works
> 
> 
> * Start with SLOW,DB (hacked vstart to bluestore_block_wal_create=false)
> * get rid of DB
>   -> fsck works, cluster works

Some DB files should be at slow device before the migrating. Hence you need to enforce some bluefs spillover (or rather force DB to put sst files to db.slow, i.e. put more data than DB volume can afford) to reproduce. See the new QA test case for the setup and repro steps.

**Comment by aclamk** [comment]
@ifed01
I have been trying to reproduce issue while I checked out your PR. Stupid me.
At least your PR got some extra testing in different conditions.

File: src/os/bluestore/BlueFS.cc [hunk 1]
```diff
@@ -3788,7 +3788,9 @@ uint8_t OriginalVolumeSelector::select_prefer_bdev(void* hint)
 void OriginalVolumeSelector::get_paths(const std::string& base, paths& res) const
 {
   res.emplace_back(base, db_total);
-  res.emplace_back(base + ".slow", slow_total);
+  res.emplace_back(base + ".slow",
+    slow_total ? slow_total : db_total); // use fake non-zero value if needed to
```
Comment (aclamk) [inline src/os/bluestore/BlueFS.cc:3792]: I think supplying "1" is better then db_total.
Works, and at least does not signal RocksDB that there is space on db.slow.

**Review by aclamk — APPROVED** [review]
Good fix.
Sorry for mistake with testing.
One optional improvement suggested.

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-09-04_01:42:42-rados-wip-kefu-testing-2021-09-04-0729-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-09-04_14:03:36-rados-wip-kefu-testing-2021-09-04-0729-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-09-05_14:04:22-rados-wip-kefu-testing-2021-09-04-0729-distro-basic-smithi/

**Pull request merged**